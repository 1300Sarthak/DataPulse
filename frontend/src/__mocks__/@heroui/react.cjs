/* eslint-disable no-undef */
const React = require('react');

module.exports = {
  Button: ({ children, ...props }) => (
    React.createElement('button', { ...props, 'data-testid': 'heroui-button' }, children)
  ),
  Spinner: (props) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-spinner' }, 'Loading...')
  ),
  Switch: (props) => (
    React.createElement('input', { type: 'checkbox', ...props, 'data-testid': 'heroui-switch' })
  ),
  Dropdown: ({ children, ...props }) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-dropdown' }, children)
  ),
  DropdownTrigger: ({ children, ...props }) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-dropdown-trigger' }, children)
  ),
  DropdownMenu: ({ children, ...props }) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-dropdown-menu' }, children)
  ),
  DropdownItem: ({ children, ...props }) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-dropdown-item' }, children)
  ),
  Card: ({ children, ...props }) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-card' }, children)
  ),
  CardBody: ({ children, ...props }) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-card-body' }, children)
  ),
  Chip: ({ children, ...props }) => (
    React.createElement('span', { ...props, 'data-testid': 'heroui-chip' }, children)
  ),
  Link: ({ children, ...props }) => (
    React.createElement('a', { ...props, 'data-testid': 'heroui-link' }, children)
  ),
  Skeleton: (props) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-skeleton' }, 'Loading...')
  ),
  Divider: (props) => React.createElement('hr', { ...props, 'data-testid': 'heroui-divider' }),
  Badge: ({ children, ...props }) => (
    React.createElement('span', { ...props, 'data-testid': 'heroui-badge' }, children)
  ),
}; 