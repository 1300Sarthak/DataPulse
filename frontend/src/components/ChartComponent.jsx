import React, { useState, useEffect } from "react";
import { Card, CardHeader, CardBody, Button, ButtonGroup } from "@heroui/react";
import {
  LineChart,
  Line,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  ResponsiveContainer,
} from "recharts";
import { useApiService } from '../services/api';

const ChartComponent = ({ symbol = "BTC", initialPeriod = "1h" }) => {
  const apiService = useApiService();
  const [data, setData] = useState([]);
  const [loading, setLoading] = useState(true);
  const [period, setPeriod] = useState(initialPeriod);
  const [error, setError] = useState(null);

  const periods = [
    { key: "1h", label: "1H" },
    { key: "24h", label: "24H" },
    { key: "7d", label: "7D" },
  ];

  // Generate mock data for development
  const generateMockData = (periodKey, basePrice = 50000) => {
    const now = new Date();
    const dataPoints = periodKey === "1h" ? 60 : periodKey === "24h" ? 24 : 7;
    const interval = periodKey === "1h" ? 1 : periodKey === "24h" ? 60 : 1440; // minutes

    return Array.from({ length: dataPoints }, (_, i) => {
      const time = new Date(now.getTime() - (dataPoints - i) * interval * 60000);
      const price = basePrice + Math.sin(i * 0.5) * 1000 + (Math.random() - 0.5) * 2000;
      return {
        time: time.toLocaleTimeString("en-US", {
          hour: "2-digit",
          minute: "2-digit",
          hour12: false,
        }),
        price: Math.max(price, 1000), // Ensure price doesn't go below $1000
      };
    });
  };

  const fetchData = async (periodKey) => {
    setLoading(true);
    setError(null);

    try {
      // Get current crypto prices and generate historical data
      const cryptoData = await apiService.getCryptoPrices();
      const currentPrice = cryptoData[symbol] || 50000;
      
      // Generate mock historical data based on current price
      const mockData = generateMockData(periodKey, currentPrice);
      setData(mockData);
    } catch (err) {
      setError("Failed to load chart data");
      console.error("Chart data fetch error:", err);
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    fetchData(period);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [symbol, period]);

  const handlePeriodChange = (newPeriod) => {
    setPeriod(newPeriod);
  };

  const formatPrice = (value) => {
    return new Intl.NumberFormat("en-US", {
      style: "currency",
      currency: "USD",
      minimumFractionDigits: 0,
      maximumFractionDigits: 0,
    }).format(value);
  };

  if (error) {
    return (
      <Card className="w-full">
        <CardBody>
          <div className="text-center text-red-500">
            <p>{error}</p>
            <Button 
              color="primary" 
              size="sm" 
              className="mt-2"
              onClick={() => fetchData(period)}
            >
              Retry
            </Button>
          </div>
        </CardBody>
      </Card>
    );
  }

  return (
    <Card className="w-full">
      <CardHeader className="flex justify-between items-center">
        <div>
          <h3 className="text-lg font-semibold">{symbol} Price History</h3>
          <p className="text-sm text-gray-500 dark:text-gray-400">
            {period === "1h" ? "Last hour" : period === "24h" ? "Last 24 hours" : "Last 7 days"}
          </p>
        </div>
        <ButtonGroup size="sm">
          {periods.map(({ key, label }) => (
            <Button
              key={key}
              color={period === key ? "primary" : "default"}
              variant={period === key ? "solid" : "bordered"}
              onClick={() => handlePeriodChange(key)}
            >
              {label}
            </Button>
          ))}
        </ButtonGroup>
      </CardHeader>
      <CardBody>
        {loading ? (
          <div className="h-64 flex items-center justify-center">
            <div className="text-gray-500 dark:text-gray-400">Loading chart...</div>
          </div>
        ) : (
          <ResponsiveContainer width="100%" height={300}>
            <LineChart data={data} margin={{ top: 5, right: 30, left: 20, bottom: 5 }}>
              <CartesianGrid strokeDasharray="3 3" stroke="#f0f0f0" />
              <XAxis
                dataKey="time"
                stroke="#888"
                fontSize={12}
                tickLine={false}
                axisLine={false}
              />
              <YAxis
                stroke="#888"
                fontSize={12}
                tickLine={false}
                axisLine={false}
                tickFormatter={formatPrice}
              />
              <Tooltip
                formatter={(value) => [formatPrice(value), "Price"]}
                labelFormatter={(label) => `Time: ${label}`}
                contentStyle={{
                  backgroundColor: "white",
                  border: "1px solid #ccc",
                  borderRadius: "4px",
                }}
              />
              <Line
                type="monotone"
                dataKey="price"
                stroke="#3b82f6"
                strokeWidth={2}
                dot={false}
                activeDot={{ r: 4, fill: "#3b82f6" }}
              />
            </LineChart>
          </ResponsiveContainer>
        )}
      </CardBody>
    </Card>
  );
};

export default ChartComponent; 