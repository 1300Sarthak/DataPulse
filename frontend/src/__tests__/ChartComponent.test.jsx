/* global describe, it, expect, beforeEach, afterEach, jest, global */
import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import ChartComponent from "../components/ChartComponent";
import '@testing-library/jest-dom';

// Mock recharts to avoid SVG rendering issues in tests
jest.mock('recharts', () => ({
  LineChart: ({ children, data }) => (
    <div data-testid="line-chart" data-points={data?.length || 0}>
      {children}
    </div>
  ),
  Line: () => <div data-testid="line" />,
  XAxis: () => <div data-testid="x-axis" />,
  YAxis: () => <div data-testid="y-axis" />,
  CartesianGrid: () => <div data-testid="cartesian-grid" />,
  Tooltip: () => <div data-testid="tooltip" />,
  ResponsiveContainer: ({ children }) => (
    <div data-testid="responsive-container">{children}</div>
  ),
}));

// Mock fetch for API calls
global.fetch = jest.fn();

describe("ChartComponent", () => {
  beforeEach(() => {
    fetch.mockClear();
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it("renders with default props (BTC, 1h)", async () => {
    render(<ChartComponent />);
    
    expect(screen.getByText("BTC Price History")).toBeInTheDocument();
    expect(screen.getByText("Last hour")).toBeInTheDocument();
    expect(screen.getByText("1H")).toBeInTheDocument();
    expect(screen.getByText("24H")).toBeInTheDocument();
    expect(screen.getByText("7D")).toBeInTheDocument();
    
    // Should show loading initially
    expect(screen.getByText("Loading chart...")).toBeInTheDocument();
    
    // Wait for mock data to load
    await waitFor(() => {
      expect(screen.queryByText("Loading chart...")).not.toBeInTheDocument();
    });
    
    // Should render chart components
    expect(screen.getByTestId("line-chart")).toBeInTheDocument();
    expect(screen.getByTestId("responsive-container")).toBeInTheDocument();
  });

  it("renders with custom symbol and period", async () => {
    render(<ChartComponent symbol="ETH" initialPeriod="24h" />);
    
    expect(screen.getByText("ETH Price History")).toBeInTheDocument();
    expect(screen.getByText("Last 24 hours")).toBeInTheDocument();
    
    await waitFor(() => {
      expect(screen.queryByText("Loading chart...")).not.toBeInTheDocument();
    });
  });

  it("changes period when buttons are clicked", async () => {
    render(<ChartComponent />);
    
    await waitFor(() => {
      expect(screen.queryByText("Loading chart...")).not.toBeInTheDocument();
    });
    
    // Click 24H button
    fireEvent.click(screen.getByText("24H"));
    expect(screen.getByText("Last 24 hours")).toBeInTheDocument();
    
    // Click 7D button
    fireEvent.click(screen.getByText("7D"));
    expect(screen.getByText("Last 7 days")).toBeInTheDocument();
  });

  it("shows error state when API fails", async () => {
    fetch.mockRejectedValueOnce(new Error("API Error"));
    
    render(<ChartComponent />);
    
    await waitFor(() => {
      expect(screen.getByText("Failed to load chart data")).toBeInTheDocument();
      expect(screen.getByText("Retry")).toBeInTheDocument();
    });
  });

  it("retries data fetch when retry button is clicked", async () => {
    fetch.mockRejectedValueOnce(new Error("API Error"));
    
    render(<ChartComponent />);
    
    await waitFor(() => {
      expect(screen.getByText("Failed to load chart data")).toBeInTheDocument();
    });
    
    // Click retry
    fireEvent.click(screen.getByText("Retry"));
    
    await waitFor(() => {
      expect(screen.queryByText("Failed to load chart data")).not.toBeInTheDocument();
    });
  });

  it("generates correct number of data points for different periods", async () => {
    render(<ChartComponent initialPeriod="1h" />);
    
    await waitFor(() => {
      const chart = screen.getByTestId("line-chart");
      expect(chart.getAttribute("data-points")).toBe("60"); // 1h = 60 points
    });
    
    // Change to 24h
    fireEvent.click(screen.getByText("24H"));
    
    await waitFor(() => {
      const chart = screen.getByTestId("line-chart");
      expect(chart.getAttribute("data-points")).toBe("24"); // 24h = 24 points
    });
    
    // Change to 7d
    fireEvent.click(screen.getByText("7D"));
    
    await waitFor(() => {
      const chart = screen.getByTestId("line-chart");
      expect(chart.getAttribute("data-points")).toBe("7"); // 7d = 7 points
    });
  });

  it("re-renders when symbol prop changes", async () => {
    const { rerender } = render(<ChartComponent symbol="BTC" />);
    
    await waitFor(() => {
      expect(screen.getByText("BTC Price History")).toBeInTheDocument();
    });
    
    // Change symbol
    rerender(<ChartComponent symbol="ETH" />);
    
    await waitFor(() => {
      expect(screen.getByText("ETH Price History")).toBeInTheDocument();
      expect(screen.queryByText("BTC Price History")).not.toBeInTheDocument();
    });
  });

  it("maintains selected period when symbol changes", async () => {
    const { rerender } = render(<ChartComponent symbol="BTC" initialPeriod="24h" />);
    
    await waitFor(() => {
      expect(screen.getByText("Last 24 hours")).toBeInTheDocument();
    });
    
    // Change symbol
    rerender(<ChartComponent symbol="ETH" initialPeriod="24h" />);
    
    await waitFor(() => {
      expect(screen.getByText("ETH Price History")).toBeInTheDocument();
      expect(screen.getByText("Last 24 hours")).toBeInTheDocument();
    });
  });
});

// Snapshot test
describe("ChartComponent Snapshot", () => {
  it("matches snapshot when loaded", async () => {
    const { container } = render(<ChartComponent />);
    
    await waitFor(() => {
      expect(screen.queryByText("Loading chart...")).not.toBeInTheDocument();
    });
    
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot in error state", async () => {
    fetch.mockRejectedValueOnce(new Error("API Error"));
    
    const { container } = render(<ChartComponent />);
    
    await waitFor(() => {
      expect(screen.getByText("Failed to load chart data")).toBeInTheDocument();
    });
    
    expect(container).toMatchSnapshot();
  });
}); 