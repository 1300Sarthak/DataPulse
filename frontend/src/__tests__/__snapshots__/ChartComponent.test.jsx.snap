// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ChartComponent Snapshot matches snapshot when loaded 1`] = `
<div>
  <div
    class="flex flex-col relative overflow-hidden h-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none w-full"
    tabindex="-1"
  >
    <div
      class="p-3 z-10 w-full shrink-0 overflow-inherit color-inherit subpixel-antialiased rounded-t-large flex justify-between items-center"
    >
      <div>
        <h3
          class="text-lg font-semibold"
        >
          BTC
           Price History
        </h3>
        <p
          class="text-sm text-gray-500"
        >
          Last hour
        </p>
      </div>
      <div
        class="inline-flex items-center justify-center h-auto"
        role="group"
      >
        <button
          class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 min-w-16 h-8 text-tiny gap-2 [&:not(:first-child):not(:last-child)]:rounded-none [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary text-primary-foreground rounded-none first:rounded-s-small last:rounded-e-small data-[hover=true]:opacity-hover"
          data-react-aria-pressable="true"
          tabindex="0"
          type="button"
        >
          1H
        </button>
        <button
          class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-3 min-w-16 h-8 text-tiny gap-2 [&:not(:first-child):not(:last-child)]:rounded-none [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground rounded-none first:rounded-s-small last:rounded-e-small [&+.border-medium.border-default]:ms-[calc(theme(borderWidth.medium)*-1)] data-[hover=true]:opacity-hover"
          data-react-aria-pressable="true"
          tabindex="0"
          type="button"
        >
          24H
        </button>
        <button
          class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-3 min-w-16 h-8 text-tiny gap-2 [&:not(:first-child):not(:last-child)]:rounded-none [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground rounded-none first:rounded-s-small last:rounded-e-small [&+.border-medium.border-default]:ms-[calc(theme(borderWidth.medium)*-1)] data-[hover=true]:opacity-hover"
          data-react-aria-pressable="true"
          tabindex="0"
          type="button"
        >
          7D
        </button>
      </div>
    </div>
    <div
      class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
    >
      <div
        data-testid="responsive-container"
      >
        <div
          data-points="60"
          data-testid="line-chart"
        >
          <div
            data-testid="cartesian-grid"
          />
          <div
            data-testid="x-axis"
          />
          <div
            data-testid="y-axis"
          />
          <div
            data-testid="tooltip"
          />
          <div
            data-testid="line"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
